### PR TITLE
Translate/add api and ui

### PR DESCRIPTION
**Pull request by benrs**
Relates to the translate API, works similarly to pinning messages, you need to select a single message and then click the icon for translating and it will translate the text. Relates to #120 for tests.

### CODE REVIEW

File: src/org/thoughtcrime/securesms/translation/TranslationTask.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+import java.io.IOException;
+import java.net.URLEncoder;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+
+public class TranslationTask extends AsyncTask<String, Void, TranslationResult> {
+    private String API_KEY = "trnsl.1.1.20180413T005640Z.e9d9285313a77d68.f90bbb8e14d7b2bd9257f87c43139b70f49023e1";
```
Comment (benrs) [inline src/org/thoughtcrime/securesms/translation/TranslationTask.java]: Going to deactivate this key and activate a new one. (And remove it in a future commit)

File: src/org/thoughtcrime/securesms/database/SmsDatabase.java [hunk 2]
```diff
@@ -501,6 +504,20 @@ private void removeReadReminder(String where, String[] arguments){
     notifyConversationListeners(threadId);
   }
 
+  public void updateMessageBody(MasterSecretUnion masterSecret, long messageId, String body) {
```
Comment (claudds) [inline src/org/thoughtcrime/securesms/database/SmsDatabase.java:509]: So this is permanently changing the body of the message in the database?  Is the original body of the message still going to be viewable?

File: test/unitTest/java/org/thoughtcrime/securesms/translation/TranslationTaskUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import okhttp3.Response;
+
+import static junit.framework.Assert.assertEquals;
+import static junit.framework.Assert.fail;
+import static org.powermock.api.mockito.PowerMockito.mock;
+import static org.powermock.api.mockito.PowerMockito.when;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({ OkHttpClient.class, Request.class, Response.class, Call.class })
+public class TranslationTaskUnitTest extends BaseUnitTest {
```
Comment (claudds) [inline test/unitTest/java/org/thoughtcrime/securesms/translation/TranslationTaskUnitTest.java:25]: Could possibly add in a unit test to the conversation fragment to make sure that the translation task is called upon handling a translation

File: src/org/thoughtcrime/securesms/database/SmsDatabase.java (hunk 2 above)
Comment (benrs) [reply src/org/thoughtcrime/securesms/database/SmsDatabase.java:509]: Yes, I set the message as "${translatedMessage} [${untranslatedMessage}]"

File: test/androidTest/java/org/thoughtcrime/securesms/espressoTests/TranslationTest.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package org.thoughtcrime.securesms.espressoTests;
```
[the remark this answers is not in the document]
Comment (benrs) [reply test/androidTest/java/org/thoughtcrime/securesms/espressoTests/TranslationTest.java]: Bad bot, I think we should exclude this type of issue.

[the remark this answers is not in the document]
Comment (claudds) [reply test/androidTest/java/org/thoughtcrime/securesms/espressoTests/TranslationTest.java]: I actually changed the name of the package in my pr

**Review by claudds — APPROVED** [review]

File: src/org/thoughtcrime/securesms/translation/TranslationTask.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        }
+        return result;
+    }
+
+    @Override
+    protected void onPostExecute(TranslationResult result) {
+        resultHandler.processTranslationResult(result, messageRecord);
+    }
+
+    public TranslationTask(OkHttpClient client, MessageRecord messageRecord) {
```
Comment (samantha-kerr) [inline src/org/thoughtcrime/securesms/translation/TranslationTask.java]: It would read easier to have the two constructors together at the top.

Comment (lori-dalkin) [reply src/org/thoughtcrime/securesms/translation/TranslationTask.java]: Agreed

**Comment by lori-dalkin** [comment]
Looks good, I'd make the few changes indicated in the comments but otherwise I'm approving!

**Review by lori-dalkin — APPROVED** [review]

**Review by samantha-kerr — APPROVED** [review]

**Comment by benrs** [comment]
Codacy doesn't see that the function is being used because it's being invoked by Whitebox (uses a string to run the method), going to merge as is.

**Pull request merged**